postgresql-plugin/src/main/java/io/cdap/plugin/postgres/PostgresErrorDetailsProvider.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright © 2024 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.postgres;
+
+import io.cdap.cdap.api.exception.ErrorCategory;
+import io.cdap.cdap.api.exception.ErrorType;
+import io.cdap.plugin.db.DBErrorDetailsProvider;
+import io.cdap.plugin.util.DBUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A custom ErrorDetailsProvider for Postgres plugins.
+ */
+public class PostgresErrorDetailsProvider extends DBErrorDetailsProvider {
+  // https://www.postgresql.org/docs/current/errcodes-appendix.html
+  private static final Map<String, ErrorType> ERROR_CODE_TO_ERROR_TYPE;
+  private static final Map<String, ErrorCategory> ERROR_CODE_TO_ERROR_CATEGORY;
+  static {
+    ERROR_CODE_TO_ERROR_TYPE = new HashMap<>();
+    ERROR_CODE_TO_ERROR_TYPE.put("01", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("02", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("08", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("0A", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("22", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("23", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("28", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("40", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("42", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("53", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("54", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("55", ErrorType.USER);
+    ERROR_CODE_TO_ERROR_TYPE.put("57", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("58", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("P0", ErrorType.SYSTEM);
+    ERROR_CODE_TO_ERROR_TYPE.put("XX", ErrorType.SYSTEM);
```

Comment:
Error Codes like `53` which represent `Insufficient Resources` returned from `PostgreSQL server` is `USER` for cdap since we do not control psql resources.

This comment applies to all such error codes classified as `SYSTEM`. PTAL, thanks.

Reply:
Does `SYSTEM` only refer to CDAP ?
I have set all to user error other than connection / postgres system/internal error .

Reply:
For cloudPostgresql will these still be user error or system given the resource here is managed.

Reply:
I discussed with PM (Deepinder) and we have decided, we will keep the `errorType` as `SYSTEM` and use `dependency: true` to indicate the error is not on CDAP side.

For `cloudsql` also `dependency: true`.

Reply:
Revert the changes, and updated dependency to true.